File: autoviz/examples/python/13_tutorial_raw_messages.py
#!/usr/bin/env python3
"""Publish nested protobufs for Autoviz Raw Messages panel.

  /fake/raw/odom   nav_msgs/Odometry
  /fake/raw/poses  geometry_msgs/PoseArray
  /fake/raw/imu    sensor_msgs/Imu

Raw Messages: select channel → full DebugString; optional Path e.g. pose.pose.pose.position.x

  /usr/bin/python3 examples/python/13_tutorial_raw_messages.py
  ./build/autonomy/bin/autoviz
"""

from __future__ import annotations

from _reexec import early_reexec_if_needed

early_reexec_if_needed()

import argparse
import math
import time

from _bootstrap import prepare_example_environment

prepare_example_environment()

import autolink
from automsgs.msgs.geometry_msgs import pose_array_pb2
from automsgs.msgs.nav_msgs import odometry_pb2
from automsgs.msgs.sensor_msgs import imu_pb2


def stamp(h, frame='map'):
    now = time.time_ns()
    h.frame_id = frame
    h.stamp.sec = now // 1_000_000_000
    h.stamp.nanosec = now % 1_000_000_000


def set_pose(pose, x, y, yaw, z=0.05):
    pose.position.x, pose.position.y, pose.position.z = x, y, z
    pose.orientation.z = math.sin(yaw * 0.5)
    pose.orientation.w = math.cos(yaw * 0.5)


def make_odom(t):
    x, y = 2.0 * math.cos(t), 1.5 * math.sin(t)
    yaw = t + math.pi / 2
    msg = odometry_pb2.Odometry()
    stamp(msg.header)
    msg.child_frame_id = 'base_link'
    stamp(msg.pose.pose.header)
    set_pose(msg.pose.pose.pose, x, y, yaw)
    msg.pose.covariance.extend([0.05 if i % 7 == 0 else 0.0 for i in range(36)])
    msg.twist.twist.linear.x = 0.8
    msg.twist.twist.angular.z = 0.4
    msg.twist.covariance.extend([0.01 if i % 7 == 0 else 0.0 for i in range(36)])
    return msg


def make_poses(t):
    msg = pose_array_pb2.PoseArray()
    stamp(msg.header)
    for i in range(5):
        a = t + i * (2 * math.pi / 5)
        set_pose(msg.poses.add(), math.cos(a), math.sin(a), a, z=0.1 * i)
    return msg


def make_imu(t):
    msg = imu_pb2.Imu()
    stamp(msg.header, 'imu_link')
    msg.orientation.z = math.sin(t * 0.5)
    msg.orientation.w = math.cos(t * 0.5)
    msg.angular_velocity.z = 0.2 * math.cos(t)
    msg.linear_acceleration.x = 0.1 * math.sin(t)
    msg.linear_acceleration.z = 9.81
    return msg


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--rate', type=float, default=10.0)
    args = ap.parse_args()

    autolink.init('raw')
    n = autolink.Node('/autoviz/raw')
    w_odom = n.create_writer('/fake/raw/odom', odometry_pb2.Odometry, qos_depth=5)
    w_poses = n.create_writer(
        '/fake/raw/poses', pose_array_pb2.PoseArray, qos_depth=5)
    w_imu = n.create_writer('/fake/raw/imu', imu_pb2.Imu, qos_depth=5)
    rate = autolink.Rate(args.rate)
    print(f'raw @ {args.rate} Hz → /fake/raw/{{odom,poses,imu}}')

    t0 = time.time()
    try:
        while not autolink.is_shutdown():
            t = time.time() - t0
            w_odom.write(make_odom(t))
            w_poses.write(make_poses(t))
            w_imu.write(make_imu(t))
            rate.sleep()
    except KeyboardInterrupt:
        pass
    finally:
        autolink.shutdown()


if __name__ == '__main__':
    main()

File: autoviz/examples/cpp/13_tutorial_raw_messages.cpp
/******************************************************************************
 * Copyright 2026 The Openbot Authors (duyongquan)
 *****************************************************************************/

#include <cmath>
#include <iostream>
#include <memory>

#include <automsgs/msgs/geometry_msgs/pose_array.pb.h>
#include <automsgs/msgs/nav_msgs/odometry.pb.h>
#include <automsgs/msgs/sensor_msgs/imu.pb.h>

#include "common/tutorial_utils.hpp"

namespace {

using autoviz::examples::SetYawPose;
using autoviz::examples::StampHeader;

std::shared_ptr<automsgs::msgs::nav_msgs::Odometry> MakeOdom(double t) {
  const double x = 2.0 * std::cos(t);
  const double y = 1.5 * std::sin(t);
  const double yaw = t + M_PI / 2.0;
  auto msg = std::make_shared<automsgs::msgs::nav_msgs::Odometry>();
  StampHeader(msg->mutable_header());
  msg->set_child_frame_id("base_link");
  auto* pwc = msg->mutable_pose();
  StampHeader(pwc->mutable_pose()->mutable_header());
  SetYawPose(pwc->mutable_pose()->mutable_pose(), x, y, yaw);
  for (int i = 0; i < 36; ++i) pwc->add_covariance(i % 7 == 0 ? 0.05 : 0.0);
  msg->mutable_twist()->mutable_twist()->mutable_linear()->set_x(0.8);
  msg->mutable_twist()->mutable_twist()->mutable_angular()->set_z(0.4);
  for (int i = 0; i < 36; ++i) {
    msg->mutable_twist()->add_covariance(i % 7 == 0 ? 0.01 : 0.0);
  }
  return msg;
}

std::shared_ptr<automsgs::msgs::geometry_msgs::PoseArray> MakePoses(double t) {
  auto msg = std::make_shared<automsgs::msgs::geometry_msgs::PoseArray>();
  StampHeader(msg->mutable_header());
  for (int i = 0; i < 5; ++i) {
    const double a = t + i * (2.0 * M_PI / 5.0);
    SetYawPose(msg->add_poses(), std::cos(a), std::sin(a), a, 0.1 * i);
  }
  return msg;
}

std::shared_ptr<automsgs::msgs::sensor_msgs::Imu> MakeImu(double t) {
  auto msg = std::make_shared<automsgs::msgs::sensor_msgs::Imu>();
  StampHeader(msg->mutable_header(), "imu_link");
  msg->mutable_orientation()->set_z(std::sin(t * 0.5));
  msg->mutable_orientation()->set_w(std::cos(t * 0.5));
  msg->mutable_angular_velocity()->set_z(0.2 * std::cos(t));
  msg->mutable_linear_acceleration()->set_x(0.1 * std::sin(t));
  msg->mutable_linear_acceleration()->set_z(9.81);
  return msg;
}

}  // namespace

using autoviz::examples::MakeWriter;
using autoviz::examples::NowSec;
using autoviz::examples::ParseRate;

int main(int argc, char** argv) {
  const double rate_hz = ParseRate(argc, argv, 10.0);

  if (!autolink::Init(argv[0])) return 1;
  auto node = autolink::CreateNode("/autoviz/raw");
  auto w_odom =
      MakeWriter<automsgs::msgs::nav_msgs::Odometry>(node, "/fake/raw/odom");
  auto w_poses = MakeWriter<automsgs::msgs::geometry_msgs::PoseArray>(
      node, "/fake/raw/poses");
  auto w_imu =
      MakeWriter<automsgs::msgs::sensor_msgs::Imu>(node, "/fake/raw/imu");

  autolink::Rate rate(rate_hz);
  std::cout << "raw @ " << rate_hz << " Hz → /fake/raw/{odom,poses,imu}\n";

  const double t0 = NowSec();
  while (autolink::OK()) {
    const double t = NowSec() - t0;
    w_odom->Write(MakeOdom(t));
    w_poses->Write(MakePoses(t));
    w_imu->Write(MakeImu(t));
    rate.Sleep();
  }
  return 0;
}
